src/System.Linq/src/System/Linq/Concat.cs

```diff
@@ -21,201 +21,83 @@ public static IEnumerable<TSource> Concat<TSource>(this IEnumerable<TSource> fir
-                {
-                    return new Concat2CollectionIterator<TSource>(firstCollection, secondCollection);
-                }
-            }
-            else
-            {
-                var firstConcat = first as ConcatIterator<TSource>;
-                if (firstConcat != null)
-                {
-                    return firstConcat.Concat(second);
-                }
-            }
-
-            return new Concat2EnumerableIterator<TSource>(first, second);
+            var firstConcat = first as ConcatIterator<TSource>;
+            return firstConcat != null ?
+                firstConcat.Concat(second) :
+                new Concat2Iterator<TSource>(first, second);
         }
 
         /// <summary>
         /// Represents the concatenation of two <see cref="IEnumerable{TSource}"/>.
         /// </summary>
         /// <typeparam name="TSource">The type of the source enumerables.</typeparam>
-        private sealed class Concat2EnumerableIterator<TSource> : ConcatIterator<TSource>
+        private sealed class Concat2Iterator<TSource> : ConcatIterator<TSource>
         {
+            /// <summary>
+            /// The first source to concatenate.
+            /// </summary>
             private readonly IEnumerable<TSource> _first;
+
+            /// <summary>
+            /// The second source to concatenate.
+            /// </summary>
             private readonly IEnumerable<TSource> _second;
 
-            internal Concat2EnumerableIterator(IEnumerable<TSource> first, IEnumerable<TSource> second)
+            /// <summary>
+            /// Initializes a new instance of the <see cref="Concat2Iterator{TSource}"/> class.
+            /// </summary>
+            /// <param name="first">The first source to concatenate.</param>
+            /// <param name="second">The second source to concatenate.</param>
+            internal Concat2Iterator(IEnumerable<TSource> first, IEnumerable<TSource> second)
             {
-                Debug.Assert(first != null && second != null);
+                Debug.Assert(first != null);
+                Debug.Assert(second != null);
+
                 _first = first;
                 _second = second;
             }
 
-            public override Iterator<TSource> Clone()
-            {
-                return new Concat2EnumerableIterator<TSource>(_first, _second);
-            }
+            public override Iterator<TSource> Clone() => new Concat2Iterator<TSource>(_first, _second);
 
             internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
             {
-                return new ConcatNEnumerableIterator<TSource>(this, next, 2);
+                // Instead of linking directly to this Concat2Iterator, we create 2 new ConcatNIterators
+                // for the first two sources and create a third one to hold the next source.
+                // This simplifies `GetEnumerable` because the nodes are of uniform type and we don't
```

Comment:
> This simplifies 'GetEnumerable'

And makes the common case of `foreach (var item in source.Concat(something).Concat(somethingElse))` more expensive, no?